staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -299,6 +321,19 @@ func (wc *watchChan) transform(e *event) (res *watch.Event) {
 	}
 
 	switch {
+	case e.isProgressNotify:
+		if wc.watcher.newFunc == nil {
+			return nil
+		}
+		object := wc.watcher.newFunc()
+		if err := wc.watcher.versioner.UpdateObject(object, uint64(e.rev)); err != nil {
+			klog.Errorf("failed to propagate object version: %v", err)
+			return nil
+		}
+		res = &watch.Event{
+			Type:   watch.Bookmark,
+			Object: object,
```

Comment:
I guess I would have expected a new bookmark type of object rather than a object that has the same type as the other objects. Is there a reason you did it this way?

Reply:
Two reasons:
Compatibility: this is exactly what we did for watch bookmarks https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md#proposal:
```
As a result, we will represent bookmark event by setting Bookmark type and Object of appropriate type with just ObjectMeta.ResourceVersion field set.
```

[The other reason is also that our client libraries (reflector, etc.) assume that incoming objects are all of the exact same type.]

Reply:
I see. I guess I didn't see/understand that originally, I wish we hadn't done it that way. I'm a little concerned unsophisticated clients might confuse this for a real object. I probably assumed "of appropriate type" meant bookmark-typed, not object-typed.

I guess, clients have to opt-in for bookmarks anyway, which kind of destroys my argument here. But it also means we could have made a bookmark type and fixed our libraries that made bad assumptions.

I guess it's too late to fix now without making clients opt-in again :(

Reply:
Yeah - I would change that in this PR at least (we may consider doing that independently from that change for watch bookmarks, but even for that I'm not sure it's really worth it).

Reply:
It'll probably never be worth it :'(

Reply:
I guess code-wise it seems like this change is a lot bigger than it needs to be because of the need to plumb the newFunc everywhere. Would it make sense to keep a bookmark-type object internally and only convert to the "real" type object right before encoding to the wire? (which I believe is here?)

Reply:
I think it could make sense in the past. But given we already did it like that for watch bookmarks, I think it's a bit too late.

> right before encoding to the wire? (which I believe is here?)

It actually isn't here. There are a number of layers that this event goes through (watchache, the fixup part (which sets selflink) etc.) so that actually wouldn't be easier codewise.

I agree that that passing newFunc is a bit unfortunate, but it's also boilerplate easy to review (and super low-risk change).